### ISSUE TITLE
Gamestates write-up

### ISSUE THREAD

**Issue by sweetsbeats** [studio-bakemono/groupong#11]
So it's coming time we start adding gamestates. I'm going to explain why we need gamestates, how we're going to do it, and then justify the method for our use. This is a bit of a wordy process, but bear with me for now.

Why:
Even with one "game state", our main file is already a manageable mess. What will happen when we try to add a main menu? Other game modes? Even at the size it is now we could justify abstracting it for readability. So while we clean up bugs, and before we start working on a menu and different game modes we should focus on creating a maintainable and scale-able method of managing multiple game states.

How:
It's a lot simpler than you'd think, so don't worry. In a game like ours, we'll only ever have one major state at a time (Menu state, playing/game, different game modes, etc.), so we can create a [finite state machine](http://gameprogrammingpatterns.com/state.html) to represent and manage game states. Our initial implementation method will involve creating a virtual gamestate object to act as a base. We'll create methods to manage the major parts of a game loop; events, update, render, etc. These virtual methods will be overwritten by actual states (so for example our "Pong" state would have all it's logic in an update method, and it's render logic in a render method).
The main file will simplify quite a lot. We will use a base gamestate pointer and use the [dynamic dispatch](https://en.wikipedia.org/wiki/Dynamic_dispatch) pattern to call the update of the current state. This is all I'll go into detail about now. It's a lot simpler in code than spoken.

So why this method:
Basically put, we could use enums, and we could use another method, however the dynamic dispatch method of handling gamestates is easy to understand, and plays nicely with finite state machines, and is a lot less error prone than say, enums.

Sometime today, I'll create a branch and start implementing this pattern. I don't expect everyone to understand everything here right away, it's a lot of words and jargon, but once we put it into code, it will become rather intuitive and I'm sure everyone will get it.
